### PR TITLE
UX-861, UX-862, UX-863/Kubevirt

### PR DESCRIPTION
**Pull request by jruddell**
Early access form of kubevirt. Made the details page a more generic page and abstracted the action logic to reuse the same list table actions for consistency. We can reuse this model for other entities like pods, deployments, services fairly easily

image

image

image

### CODE REVIEW

**Review by knnguy — APPROVED** [review]

File: src/app/plugins/kubernetes/components/virtual-machines/add.tsx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      VirtualMachineCreateTypes.AddNew
+    )
+  }
+  return VirtualMachineCreateTypes.AddNew
+}
+const formTitleByType = {
+  [VirtualMachineCreateTypes.AddNew]: 'Create a New VM Instance',
+  [VirtualMachineCreateTypes.ImportURL]: 'Add a VM - Import an Image from URL',
+  [VirtualMachineCreateTypes.ImportDisk]: 'Add a VM - Import an Image on a Disk',
+  [VirtualMachineCreateTypes.ClonePVC]: 'Add a VM - Cloning an Exisiting PVC',
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/virtual-machines/add.tsx:51]: Typo 'Existing'

File: src/app/plugins/kubernetes/components/virtual-machines/actions.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  [VMVolumeTypes.ContainerDisk]: null,
+  [VMVolumeTypes.EmptyDisk]: null,
+  [VMVolumeTypes.HostDisk]: null,
+  [VMVolumeTypes.ConfigMap]: null,
+  [VMVolumeTypes.Secret]: null,
+  [VMVolumeTypes.ServiceAccount]: null,
+}
+export const findVolumeSpecPromises = (clusterId, namespace, volumes: Volume[] = []) => {
+  const promises = []
+  volumes.forEach((volume) => {
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/virtual-machines/actions.ts:31]: could it make more sense to do a map and then a filter?

File: src/app/plugins/kubernetes/components/virtual-machines/list-page.tsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
+import React, { useCallback } from 'react'
+import { makeStyles } from '@material-ui/styles'
+import createCRUDComponents from 'core/helpers/createCRUDComponents'
+import ClusterPicklist from 'k8s/components/common/ClusterPicklist'
+import useDataLoader from 'core/hooks/useDataLoader'
+import { createUsePrefParamsHook } from 'core/hooks/useParams'
+import { listTablePrefs, allKey } from 'app/constants'
+import { pick } from 'ramda'
+import NamespacePicklist from 'k8s/components/common/NamespacePicklist'
+// import { DateAndTime } from 'core/components/listTable/cells/DateCell'
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/virtual-machines/list-page.tsx:10]: extra (already imported below)

File: src/app/plugins/kubernetes/components/virtual-machines/list-page.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import { routes } from 'core/utils/routes'
+// import { trackEvent } from 'utils/tracking'
+// import SimpleLink from 'core/components/SimpleLink'
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/virtual-machines/list-page.tsx:13]: extra

File: src/app/plugins/kubernetes/components/virtual-machines/power-vm-dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        vm_name: virtualMachine?.name,
+        vm_namespace: virtualMachine?.namespace,
+      })
+    },
+  )
+  const title = `Permanently delete VM "${virtualMachine?.name}"?`
+  const handleAction = useCallback(async () => {
+    await powerVirtualMachine(virtualMachine)
+    onClose()
+    // history.push(routes.virtualMachines.list.path())
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/virtual-machines/power-vm-dialog.tsx:28]: still needed?

**Review by kevinylee1991 — APPROVED** [review]
Noticed a number of commented code I didn't mark, are they for future building blocks?

File: src/app/plugins/kubernetes/components/virtual-machines/power-vm-dialog.tsx (hunk 5 above)
Comment (jruddell) [reply src/app/plugins/kubernetes/components/virtual-machines/power-vm-dialog.tsx:28]: pretty much all of the commented out stuff is needed but just disabled for now

File: src/app/plugins/kubernetes/components/virtual-machines/actions.ts (hunk 2 above)
Comment (jruddell) [reply src/app/plugins/kubernetes/components/virtual-machines/actions.ts:31]: theres a lot of mappings going on so i wanted it to be explicit adds. we could do that but its a little excessive

**Pull request merged**